### PR TITLE
BUG: Fix mismatch in definition and declaration for a couple functions (see #27719)

### PR DESCRIPTION
**Pull request by snitish**
Addresses #27719
```

- NumPyOS_ascii_tolower declaration conflicting its definition. The declaration says that the input parameter is a char while the definition says it is an `int.

- arr_interp declaration has one parameter more than the definition.

- arr_interp_complex declaration has one parameter more than the definition.

```

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
How do you reproduce the originally-reported build error? For me, the build succeeds on x86_64 Linux with gcc `14.2.0` and on latest NumPy `main` via i.e., `spin build -j 16 -- -Db_lto=true` and even on this branch there is still some noise in the build:

<details>

```
[496/517] Linking target numpy/linalg/_umath_linalg.cpython-311-x86_64-linux-gnu.so
../numpy/_core/include/numpy/npy_math.h:509:7: warning: type of 'npy_cabsf' does not match original declaration [-Wlto-type-mismatch]
  509 | float npy_cabsf(npy_cfloat z);
      |       ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: type mismatch in parameter 1
 1758 | npy_@kind@@c@(@ctype@ z)
      | ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: type 'complex float' should match type 'struct npy_cfloat'
../numpy/_core/include/numpy/npy_common.h:371:3: note: the incompatible type is defined here
  371 | } npy_cfloat;
      |   ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: 'npy_cabsf' was previously declared here
 1758 | npy_@kind@@c@(@ctype@ z)
      | ^
../numpy/_core/include/numpy/npy_math.h:481:8: warning: type of 'npy_cabs' does not match original declaration [-Wlto-type-mismatch]
  481 | double npy_cabs(npy_cdouble z);
      |        ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: type mismatch in parameter 1
 1758 | npy_@kind@@c@(@ctype@ z)
      | ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: type 'complex double' should match type 'struct npy_cdouble'
../numpy/_core/include/numpy/npy_common.h:366:3: note: the incompatible type is defined here
  366 | } npy_cdouble;
      |   ^
../numpy/_core/src/npymath/npy_math_complex.c.src:1758:1: note: 'npy_cabs' was previously declared here
 1758 | npy_@kind@@c@(@ctype@ z)
      | ^
[517/517] Linking target numpy/_core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so
```

</details>

Maybe there is a bit less build noise on this branch, but where are the original hard build errors coming from? Not a default build right?

**Comment by snitish** [comment]
@tylerjereddy I didn't get any hard errors but did see warnings that my change eliminates. The original reporter might have enabled errors for lto-type-mismatch warnings. I did verify that the below command fails:
```
spin build --clean -j 16 -- -Db_lto=true --werror
```
with errors (among others):
```
FAILED: numpy/_core/_multiarray_umath.cpython-312-x86_64-linux-gnu.so 
c++ @numpy/_core/_multiarray_umath.cpython-312-x86_64-linux-gnu.so.rsp
../numpy/_core/src/common/numpyos.h:54:1: error: type of ‘NumPyOS_ascii_tolower’ does not match original declaration [-Werror=lto-type-mismatch]
   54 | NumPyOS_ascii_tolower(char c);
      | ^
../numpy/_core/src/common/numpyos.c:419:1: note: type mismatch in parameter 1
  419 | NumPyOS_ascii_tolower(int c)
      | ^
../numpy/_core/src/common/numpyos.c:419:1: note: type ‘int’ should match type ‘char’
../numpy/_core/src/common/numpyos.c:419:1: note: ‘NumPyOS_ascii_tolower’ was previously declared here
../numpy/_core/src/multiarray/compiled_base.h:15:1: error: type of ‘arr_interp_complex’ does not match original declaration [-Werror=lto-type-mismatch]
   15 | arr_interp_complex(PyObject *, PyObject *const *, Py_ssize_t, PyObject *, PyObject *);
      | ^
../numpy/_core/src/multiarray/compiled_base.c:716:1: note: type mismatch in parameter 5
  716 | arr_interp_complex(PyObject *NPY_UNUSED(self), PyObject *const *args, Py_ssize_t len_args,
      | ^
../numpy/_core/src/multiarray/compiled_base.c:716:1: note: ‘arr_interp_complex’ was previously declared here
../numpy/_core/src/multiarray/compiled_base.h:13:1: error: type of ‘arr_interp’ does not match original declaration [-Werror=lto-type-mismatch]
   13 | arr_interp(PyObject *, PyObject *const *, Py_ssize_t, PyObject *, PyObject *);
      | ^
../numpy/_core/src/multiarray/compiled_base.c:546:1: note: type mismatch in parameter 5
  546 | arr_interp(PyObject *NPY_UNUSED(self), PyObject *const *args, Py_ssize_t len_args,
      | ^
../numpy/_core/src/multiarray/compiled_base.c:546:1: note: ‘arr_interp’ was previously declared here
```

**Comment by tylerjereddy** [comment]
`spin build --clean -j 16 -- -Db_lto=true --werror` fails on this branch too though. I'm basically looking for an easy way to quantify the improvement in the branch here.

This seems to be clear enough--the following shows 2 warnings on this branch and 5 on `main`:

```
spin build -j 16 -- -Db_lto=true >& log.txt
cat log.txt | grep -E -i "warning" | wc -l
```

The diff seems simple enough and the CI fails are unrelated so this patch is probably safe, though I'll likely let a more active core dev make the final call.

One minor note--for commit messages we usually use prefixes as explained here: https://numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

Anyway, this patch seems useful I think.

**Comment by mattip** [comment]
hmm. The arr_interp/arr_interp_complex additional arguments came in from #23018. I guess they were harmless since the vectorcall convention would only ever fill in 4 arguments.

I think for `NumPyOS_ascii_tolower` the situation is more complicated. The argument passed into the function is always a `char`, so I think more refactoring is needed to make the declaration and usage match `char NumPyOS_ascii_tolower(char);`

File: numpy/_core/src/common/numpyos.c [hunk 1]
```diff
@@ -416,7 +416,7 @@ NumPyOS_ascii_isupper(char c)
  * Same as tolower under C locale
  */
 NPY_NO_EXPORT int
-NumPyOS_ascii_tolower(int c)
+NumPyOS_ascii_tolower(char c)
```
Comment (mattip) [inline numpy/_core/src/common/numpyos.c:419]: The definition is correct, it is meant to replace `int tolower(int)`. Please change the declaration.

Comment (snitish) [reply numpy/_core/src/common/numpyos.c]: Thanks Matti, updated the declaration.

**Comment by mattip** [comment]
Thanks @snitish

**Pull request merged**